src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -736,6 +774,54 @@ private void CopyChangedDynamicValues(TStructuralType targetEntity)
                 targetEntity);
         }
 
+        /// <summary>
+        /// Copies changed nested properties and leaves the unchanged properties
+        /// </summary>
+        /// <param name="original">The structural object</param>
+        private void CopyChangedNestedProperties(TStructuralType original)
+        {
+            // For nested resources.
+            foreach (string nestedResourceName in _deltaNestedResources.Keys)
+            {
+                // Patch for each nested resource changed under this TStructuralType.
+                dynamic deltaNestedResource = _deltaNestedResources[nestedResourceName];
+                dynamic originalNestedResource = null;
+
+                if (!TryGetPropertyRef(original, nestedResourceName, out originalNestedResource))
+                {
+                    throw Error.Argument(nestedResourceName, SRResources.DeltaNestedResourceNameNotFound,
+                        nestedResourceName, original.GetType());
+                }
+
+                if (originalNestedResource == null)
+                {
+                    // When patching original target of null value, directly set nested resource.
+                    dynamic deltaObject = _deltaNestedResources[nestedResourceName];
+                    dynamic instance = deltaObject.GetInstance();
+
+                    // Recursively patch up the instance with the nested resources.
+                    deltaObject.CopyChangedValues(instance);
+
+                    _allProperties[nestedResourceName].SetValue(original, instance);
+                }
+                else
+                {
+                    // Recursively patch the subtree.                    
+                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+
+                    Type newType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && newType != originalType)
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
```

Comment:
Someone needs to scroll horizontally a lot to view the entire line. Considering breaking it into multiple lines